File: python/med/q21.py
class WaterBucketQueue:
    def __init__(self):
        self.pouring_bucket = []
        self.delivery_bucket = []

    def pour_water(self):
        while self.pouring_bucket:
            self.delivery_bucket.append(self.pouring_bucket.pop())

    def fill_water(self, water):
        self.delivery_bucket.append(water)

    def deliver_water(self):
        if not self.delivery_bucket:
            if not self.pouring_bucket:
                raise RuntimeError("Buckets are empty!")
            self.pour_water()
        return self.delivery_bucket.pop()

    def peek_water(self):
        if not self.pouring_bucket and not self.delivery_bucket:
            raise RuntimeError("Buckets are empty!")
        if not self.delivery_bucket:
            self.pour_water()
        return self.delivery_bucket[-1]

    def is_empty(self):
        return not self.pouring_bucket

# Test Case
bucket_queue = WaterBucketQueue()
bucket_queue.fill_water(10)
bucket_queue.fill_water(20)
print("Peek Water:", bucket_queue.peek_water())  # Expected: 10
print("Deliver Water:", bucket_queue.deliver_water())  # Expected: 10
print("Is Empty:", bucket_queue.is_empty())  # Expected: False

File: js/med/q21.js
class WaterBucketQueue {
    constructor() {
        this.pouringBucket = [];
        this.deliveryBucket = [];
    }

    pourWater() {
        while (this.pouringBucket.length > 0) {
            this.deliveryBucket.push(this.pouringBucket.pop());
        }
    }

    fillWater(water) {
        this.deliveryBucket.push(water);
    }

    deliverWater() {
        if (this.deliveryBucket.length === 0) {
            if (this.pouringBucket.length === 0) {
                throw new Error("Buckets are dry!");
            }
        }
        return this.deliveryBucket.pop();
    }

    peekWater() {
        if (this.pouringBucket.length === 0 && this.deliveryBucket.length === 0) {
            throw new Error("Buckets are dry!");
        }
        return this.deliveryBucket[this.deliveryBucket.length - 1];
    }

    isEmpty() {
        return this.pouringBucket.length === 0;
    }
}

// Test Case
const bucketQueue = new WaterBucketQueue();
bucketQueue.fillWater(10);
bucketQueue.fillWater(20);
console.log("Peek Water:", bucketQueue.peekWater()); // Expected: 10
console.log("Deliver Water:", bucketQueue.deliverWater()); // Expected: 10
console.log("Is Empty:", bucketQueue.isEmpty()); // Expected: false
